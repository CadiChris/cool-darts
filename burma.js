import { INSCRIRE_JOUEUR, DEMARRER_PARTIE, VOLEE } from "./burma.actions";

const burma = (state = STATE_INITIAL, action) => {
  switch (action.type) {
    case INSCRIRE_JOUEUR:
      return {
        ...state,
        joueurs: [...state.joueurs, action.payload.nomDuJoueur]
      }

    case DEMARRER_PARTIE:
      return {
        ...state,
        lanceur: state.joueurs[0],
        scores: state.joueurs.map(scoreVierge),
        phase: 'EN_COURS',
        chiffreCourant: 15
      }

    case VOLEE:
      const prochainLanceur = state.joueurs[(state.joueurs.indexOf(state.lanceur) + 1) % state.joueurs.length];
      const changementDeChiffre = prochainLanceur === state.joueurs[0]
      return {
        ...state,
        scores: [
          ...state.scores.map((s) =>
              noterUneVolee(state.chiffreCourant, action.payload.nombreDeTouches, s, state.lanceur))
        ],
        lanceur: prochainLanceur,
        chiffreCourant: changementDeChiffre ? CHIFFRES_DU_BURMA[CHIFFRES_DU_BURMA.indexOf(state.chiffreCourant) + 1] : state.chiffreCourant
      }

    default:
      return state
  }
}

const STATE_INITIAL = {
  joueurs: [],
  lanceur: undefined,
  scores: [],
  phase: "INSCRIPTION"
}

const POINTS_INITIAUX = 40;
const scoreVierge = (joueur) => ({
  joueur,
  points: POINTS_INITIAUX,
  touches: {}
})

const noterUneVolee = (chiffre, nombreDeTouches, score, lanceur) => {
  const joueurNonConcerne = score.joueur !== lanceur
  if (joueurNonConcerne)
    return { ...score }

  if (nombreDeTouches !== 0)
    return {
      ...score,
      points: score.points + chiffre * nombreDeTouches,
      touches: {
        ...score.touches,
        [chiffre]: nombreDeTouches
      }
    }
}

const CHIFFRES_DU_BURMA = [15, 16, 'D', 17, 18, 'T', 19, 20, 'B']

export {
  burma
}