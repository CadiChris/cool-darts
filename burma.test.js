import { burma } from './burma'

it('ne fait rien', () => {
  expect(burma()).toEqual({})
})