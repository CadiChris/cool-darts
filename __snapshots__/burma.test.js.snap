// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`démarre la partie 1`] = `
Object {
  "chiffreCourant": 15,
  "joueurs": Array [
    "J1",
    "J2",
  ],
  "lanceur": "J1",
  "phase": "EN_COURS",
  "scores": Array [
    Object {
      "joueur": "J1",
      "points": 40,
      "touches": Object {},
    },
    Object {
      "joueur": "J2",
      "points": 40,
      "touches": Object {},
    },
  ],
}
`;

exports[`inscrit un joueur 1`] = `
Object {
  "joueurs": Array [
    "J1",
  ],
  "lanceur": undefined,
  "phase": "INSCRIPTION",
  "scores": Array [],
}
`;

exports[`note une volée 1`] = `
Object {
  "chiffreCourant": 15,
  "joueurs": Array [
    "J1",
    "J2",
  ],
  "lanceur": "J2",
  "phase": "EN_COURS",
  "scores": Array [
    Object {
      "joueur": "J1",
      "points": 55,
      "touches": Object {
        "15": 1,
      },
    },
    Object {
      "joueur": "J2",
      "points": 40,
      "touches": Object {},
    },
  ],
}
`;

exports[`note une volée 2`] = `
Object {
  "chiffreCourant": 16,
  "joueurs": Array [
    "J1",
    "J2",
  ],
  "lanceur": "J1",
  "phase": "EN_COURS",
  "scores": Array [
    Object {
      "joueur": "J1",
      "points": 55,
      "touches": Object {
        "15": 1,
      },
    },
    Object {
      "joueur": "J2",
      "points": 70,
      "touches": Object {
        "15": 2,
      },
    },
  ],
}
`;

exports[`retourne le state initial 1`] = `
Object {
  "joueurs": Array [],
  "lanceur": undefined,
  "phase": "INSCRIPTION",
  "scores": Array [],
}
`;
